C++/Toolbox/Option/Option.h

```diff
@@ -0,0 +1,127 @@
+#ifndef _OPTION_H_
+#define _OPTION_H_
+
+#include <functional>
+#include <memory>
+#include <type_traits>
+
+namespace Toolbox {
+namespace Option {
+
+  /* The fundamental class describing an Option. Users only collect the Opt<T> as a result and do
+     do not instantiate this class directly. The static methods Some, None, and Option have 
+     standalone function aliases for convenience.
+
+     The underlying data structure is a std::unique_ptr. An Opt<T> is move only. 
+
+     The fundamental operations are forEach, map, and filter. `forEach` simply allows access to the
+     underlying data if it is Some. `map` and `filter` allow access and modification of the Opt<T>.
+  */ 
+  template<typename T>
+  class Opt
+  {
+  public:
+    // forward arguments to construct a Some<T>
+    template<typename... Args>
+    static Opt Some(Args... args)
+    {
+      return Option::Opt<T>(std::move(std::make_unique<T>(std::forward<Args>(args)...)));
+    }
+
+    // Create a None<T>
+    static Opt None()
+    {
+      return Option::Opt<T>();
+    }
+
+    // Create an Option from a pointer
+    // None<T> if is nullptr
+    // Else Some<T>
+    static Opt Option(T* t)
+    {
+      return Option::Opt<T>(std::move(std::unique_ptr<T>(t)));
+    }
+
+  private:
+    // Hidden constructors - one can't instantiate an Opt directly, only the class can
+    Opt(void) = default;
+    Opt(std::unique_ptr<T> p) : m_pData(std::move(p)) {};
+
+  public:
+    // No copy constructor available, explicitly delete so that it so intent is clear
+    template<typename T>
+    Opt(const Opt&) = delete;
+
+  public:
+    // operator bool for quick check
+    operator bool() const {
+      return (m_pData != nullptr);
+    };
+
+  public:
+    // forEach - read the value if it exists
+    void forEach(std::function<void(const T&)> forEachFunction) {
```

Comment:
Just reiterating my comments in another discussion, but forEach seems a bit weird in a 1 or 0 situation like option.  One option I brought up was the use of the parenthesis operator, although there could simply be a better worded function name that better represents the operation.

Reply:
Yeah I'm actually intrigued by the parens operator. I think it could be elegant. `forEach` is convention in Scala...but we are writing C++ and we are rolling our own Option. 

This could be the implementation:

```
void operator() (std::function<void(const T& t)> function) {
  if (m_pData)
  function(*m_pData.get());
}
```
and usage:
```
auto widgetOption = Toolbox::Option::Some<Widget>(10);
widgetOption([](const Widget& widget){
  // do something with widget
});
```

Reply:
I am personally a fan of this approach but I am still open to a named method that better defines the operation we are performing.  Something along the lines of: do, process, operate (arguably this is equivalent semantically to the paren operator), execute, etc.  I am potentially a fan of 'process'.

Reply:
I'm changing it to `ifPresent`. This is clear and is a method used in Java's Optional type.